Source/WebCore/css/StyleColor.h

```diff
@@ -31,31 +31,86 @@
 
 #pragma once
 
-#include "CSSValueKeywords.h"
 #include "Color.h"
+#include "ColorInterpolationMethod.h"
+#include "CSSValueKeywords.h"
+#include <vector>
 #include <wtf/OptionSet.h>
 
 namespace WebCore {
 
+class ColorMixComponent;
+class CSSPrimitiveValue;
+class StyleColor;
+
 enum class StyleColorOptions : uint8_t {
     ForVisitedLink = 1 << 0,
     UseSystemAppearance = 1 << 1,
     UseDarkAppearance = 1 << 2,
     UseElevatedUserInterfaceLevel = 1 << 3
 };
 
-struct StyleColor {
+struct CurrentColor {
+    bool operator==(CurrentColor const&) const = default;
+    static StyleColor make();
+};
+
+struct SystemColor {
+    bool operator==(SystemColor const&) const = default;
+};
+struct ColorContext {
+    std::optional<Color> currentColor = std::nullopt;
+};
+
+using ColorKind = std::variant<Color, CurrentColor, SystemColor>;
+
+class StyleColor {
+public:
     static Color colorFromKeyword(CSSValueID, OptionSet<StyleColorOptions>);
+    static Color colorFromAbsoluteKeyword(CSSValueID);
     static bool isAbsoluteColorKeyword(CSSValueID);
+    static bool isCurrentColorKeyword(CSSValueID);
+    static bool isCurrentColor(CSSPrimitiveValue const&);
     WEBCORE_EXPORT static bool isSystemColorKeyword(CSSValueID);
 
     enum class CSSColorType : uint8_t {
         Absolute = 1 << 0,
         Current = 1 << 1,
         System = 1 << 2,
     };
+
     // https://drafts.csswg.org/css-color-4/#typedef-color
     static bool isColorKeyword(CSSValueID, OptionSet<CSSColorType> = { CSSColorType::Absolute, CSSColorType::Current, CSSColorType::System });
+
+    StyleColor();
+    StyleColor(Color const&);
+    StyleColor(ColorKind const&);
+    StyleColor(StyleColor const&) = default;
+    StyleColor& operator=(StyleColor const&) = default;
+    bool operator==(StyleColor const&) const = default;
+
+    StyleColor(SRGBA<uint8_t> const&);
+
+    bool isCurrentColor() const;
+    bool isSystemColor() const;
+    bool isAbsoluteColor() const;
+    bool isColorMix() const;
+    Color const& absoluteColor() const;
+    void setInheritedColor(const Color&);
+    std::optional<Color> inheritedColor() const;
+
+    WEBCORE_EXPORT Color resolveColor(ColorContext = { }) const;
```

Comment:
How would callers know when to pass a ColorContext and when not (since it is optional)?

Reply:
I've added a comment to explain the purpose of this. Normally you should always provide a context, but the state of the codebase makes it impossible sometimes.
Maybe we should remove the optional value and make it explicit that you pass an empty context ?